builtin/audit/socket/backend.go

```diff
@@ -428,10 +428,29 @@ func (b *Backend) configureSinkNode(name string, address string, format string,
 		return fmt.Errorf("%s: error creating socket sink node: %w", op, err)
 	}
 
-	sinkNode := &audit.SinkWrapper{Name: name, Sink: n}
+	// Wrap the sink node with metrics middleware
+	sinkMetricTimer, err := audit.NewSinkMetricTimer(name, n)
+	if err != nil {
+		return fmt.Errorf("%s: unable to add timing metrics to sink for path %q: %w", op, name, err)
+	}
+
+	// Decide what kind of labels we want and wrap the sink node inside a metrics counter.
+	var metricLabeler event.Labeler
+	switch {
+	case b.fallback:
+		metricLabeler = &audit.MetricLabelerAuditFallback{}
+	default:
+		metricLabeler = &audit.MetricLabelerAuditSink{}
+	}
+
+	sinkMetricCounter, err := event.NewMetricsCounter(name, sinkMetricTimer, metricLabeler)
+	if err != nil {
+		return fmt.Errorf("%s: unable to add counting metrics to sink for path %q: %w", op, name, err)
+	}
 
 	b.nodeIDList = append(b.nodeIDList, sinkNodeID)
-	b.nodeMap[sinkNodeID] = sinkNode
+	b.nodeMap[sinkNodeID] = sinkMetricCounter
```

Comment:
🤔 I see that `HasFiltering` below checks if `b.nodeMap` can be nil. With that said, it seems like we could panic here due to a nil pointer dereference. If that is actually possible in this code path, it would be prudent to guard against it.

This logic also exists within `configureSinkNode` in the `file` and `syslog` audit backends.

Reply:
'kind of' 😄 (and good spot too)

`HasFiltering` is an exported method from the package so it felt kinder to others to make sure that wouldn't explode, but these `configure___` methods were pulled out of the `Factory` methods to help clean things up and make them more testable, so the intention is really that they're just called within the `file`, `socket` and `syslog` packages which are small (unlike our infamous `vault` package). 

We've got a similar thing going on in `configureFormatterNode` for example, and I think things might start to get a bit more interesting in trying to decide if an error should come back from each method or it should just `make` the map and carry on. Because a filter node is the first node in a pipeline, so maybe it can 'fix' things, but a sink node is the last. 

I'll have a ponder, because in a way I don't know if things exploding there aren't as bad as the `HasFiltering` one which might be called elsewhere. 🤔

Reply:
I'm tempted to leave this one for now, as the existing `Factory` funcs are like our constructors, so if you're manually `&file.Backend{}` etc. then you're in a bit of trouble.